src/rgw/rgw_group.cc

```diff
@@ -0,0 +1,483 @@
+#include <errno.h>
+#include <ctime>
+#include <regex>
+
+#include "common/errno.h"
+#include "common/Formatter.h"
+#include "common/ceph_json.h"
+#include "common/ceph_time.h"
+#include "rgw_rados.h"
+
+#include "include/types.h"
+#include "rgw_string.h"
+
+#include "rgw_common.h"
+#include "rgw_tools.h"
+#include "rgw_group.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+using namespace std;
+
+const string RGWGroup::group_name_oid_prefix = "group_names.";
+const string RGWGroup::group_oid_prefix = "group.";
+const string RGWGroup::group_path_oid_prefix = "group_paths.";
+const string RGWGroup::group_arn_prefix = "arn:aws:iam::";
+
+int RGWGroup::store_info(bool exclusive)
+{
+  string oid = get_info_oid_prefix() + id;
+
+  bufferlist bl;
+  ::encode(*this, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().groups_pool, oid,
+                bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWGroup::store_name(bool exclusive)
+{
+  RGWNameToId nameToId;
+  nameToId.obj_id = id;
+
+  string oid = tenant + get_names_oid_prefix() + name;
+
+  bufferlist bl;
+  ::encode(nameToId, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().groups_pool, oid,
+              bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWGroup::store_path(bool exclusive)
+{
+  string oid = tenant + get_path_oid_prefix() + path + get_info_oid_prefix() + id;
+
+  return rgw_put_system_obj(store, store->get_zone_params().groups_pool, oid,
+              NULL, 0, exclusive, NULL, real_time(), NULL);
+}
+
+int RGWGroup::create(bool exclusive)
+{
+  int ret;
+
+  if (! validate_input()) {
+    return -EINVAL;
+  }
+
+  /* check to see the name is not used */
+  ret = read_id(name, tenant, id);
+  if (exclusive && ret == 0) {
+    ldout(cct, 0) << "ERROR: name " << name << " already in use for group id "
+                    << id << dendl;
+    return -EEXIST;
+  } else if ( ret < 0 && ret != -ENOENT) {
+    ldout(cct, 0) << "failed reading group id  " << id << ": "
+                  << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  /* create unique id */
+  uuid_d new_uuid;
+  char uuid_str[37];
+  new_uuid.generate_random();
+  new_uuid.print(uuid_str);
+  id = uuid_str;
+
+  //arn
+  arn = group_arn_prefix + tenant + ":group" + path + name;
+
+  // Creation time
+  real_clock::time_point t = real_clock::now();
+
+  struct timeval tv;
+  real_clock::to_timeval(t, tv);
+
+  char buf[30];
+  struct tm result;
+  gmtime_r(&tv.tv_sec, &result);
```

Comment:
is using the boost::str(boost::format( idiom an option here (avoiding need for small buffer edits and ansi calls)?